runtime/doc/vim9.txt

```diff
@@ -1698,6 +1698,140 @@ argument type checking: >
 Types are checked for most builtin functions to make it easier to spot
 mistakes.
 
+Categories of variables, defaults and null handling ~
+				*variable-categories* *null-variables*
+There are categories of variables:
+	primitive	number, float, boolean
+	container	string, blob, list, dict
+	specialized	function, job, channel, user-defined-object
+
+When declaring a variable without an initializer, an explicit type must be
+provided. Each category has different default initialization semantics. Here's
+an example for each category: >
+	var num: number		# primitives default to a 0 equivalent
+	var cont: list<string>	# containers default to an empty container
+	var spec: job		# specialized variables default to null
+<
+Vim does not have a familiar null value; it has various null_<type> predefined
+values, for example |null_string|, |null_list|, |null_job|. Primitives do not
+have a null_<type>. The typical use cases for null_<type> are:
+- to `clear a variable` and release its resources;
+- as a `default for a parameter` in a function definition.
+
+For a specialized variable, like `job`, null_<type> is used to clear the
+resources. For a container variable, resources can also be cleared by
+assigning an empty container to the variable. For example: >
+	var j: job = job_start(...)
+	# job does its work
+	j = null_job
+	var l: list<any>
+	# ... add lots of stuff to list
+	l = []  # clear the variable and release container resources
+Using the empty container, rather than null_<type>, may avoid null
+complications as described in |null-anomalies|.
+
+The initialization semantics of container and specialized variables differ: >
+	var l1: list<string>		    # empty container
+	var l1: list<string> = []	    # empty container
+	var l2: list<string> = null_list    # null container
+Specialized variables default to null.
+These job initializations are equivalent and indistinguishable: >
+	var j1: job
+	var j2: job = null_job
+	var j3 = null_job
+Initializing a list or dict with null_<type>, without specifying
+the item type, defaults to any: >
+	var d = null_dict
+	echo typename(d)	# output: "dict<any>"
+|vim9-func-declaration| are particularly unique.
+						*null-compare*
+For familiar null compare semantics, do not use null_<type> in a comparison,
+then it easy to distinguish an empty container from a null container: >
+	vim9script
+	def F(arg: list<string> = null_list)
+	    if arg == null
+	       echo "null"
+	    else
+		echo printf("not null, %sempty", empty(arg) ? '' : 'not ')
+	    endif
+	enddef
+	F()		# output: "null"
+	F(null_list)	# output: "null"
+	F([])		# output: "not null, empty"
+	F([0])		# output: "not null, not empty"
```

Comment:
1, though this work, and `is` is the right way to compare, but i think it is quite easy to write wrong or problem code.